File: hms-backend/src/main/java/com/product/hms/api/RatingApi.java
package com.product.hms.api;

import com.product.hms.dto.response.RatingSummaryResponse;
import com.product.hms.service.RatingService;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;

@RestController
@RequestMapping("/api/ratings")
@RequiredArgsConstructor
public class RatingApi {

    private final RatingService ratingService;

    @GetMapping("/room-class/{roomClassId}")
    public ResponseEntity<Map<String, Object>> getRatingsByRoomClass(
            @PathVariable Long roomClassId,
            @RequestParam(required = false) Integer rating,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "5") int size) {
        
        Pageable pageable = PageRequest.of(page, size);
        Page<RatingSummaryResponse> feedbackPage = ratingService.getRatingsByRoomClass(roomClassId, rating, pageable);

        RatingSummaryResponse feedbackData = feedbackPage.getContent().isEmpty()
                ? new RatingSummaryResponse()
                : feedbackPage.getContent().get(0);

        Map<String, Object> response = new HashMap<>();
        response.put("averageRating", feedbackData.getAverageRating() != null ? feedbackData.getAverageRating() : 0.0);
        response.put("totalReviews", feedbackData.getTotalReviews() != null ? feedbackData.getTotalReviews() : 0L);
        response.put("ratingDistribution", feedbackData.getRatingDistribution());
        response.put("content", feedbackData.getContent());
        response.put("currentPage", feedbackPage.getNumber());
        response.put("totalPages", feedbackPage.getTotalPages());
        response.put("totalItems", feedbackPage.getTotalElements());
        response.put("pageSize", feedbackPage.getSize());
        response.put("hasNext", feedbackPage.hasNext());
        response.put("hasPrevious", feedbackPage.hasPrevious());

        return ResponseEntity.ok(response);
    }
}

File: hms-frontend/src/apis/ratingApi.js
import axios from 'axios';

const API_URL = 'http://localhost:8080/api/ratings';

export const getRoomClassRatings = async (roomClassId, page = 0, size = 5, rating = null) => {
    try {
        const params = {
            page,
            size
        };

        if (rating !== null && rating !== 'all') {
            params.rating = rating;
        }

        const response = await axios.get(`${API_URL}/room-class/${roomClassId}`, { params });
        return response.data;
    } catch (error) {
        console.error('Error fetching ratings:', error);
        throw error;
    }
};

export const getLatestRatings = async (page = 0, size = 3) => {
    try {
        const response = await axios.get(`${API_URL}/latest`, {
            params: { page, size }
        });
        return response.data;
    } catch (error) {
        console.error('Error fetching latest ratings:', error);
        throw error;
    }
};
